Remove pinnable buffer from SslStream

Drawaes opened the pull request:
In testing underload and with different threading and programming models that the Pinnable buffer is either not percievably faster or significantly slower than just using the shared array pool. By removing this we remove a significant amount of code from this class and pave the way for other refactors.

In the current state I see a small performance bump in SslStream underload on Unix on a NUMA based system. Maybe if he could be so kind @benaadams or someone with windows benchmarking could confirm the same on Windows. I see from 149k -> 160k over a number of test runs, it's just outside the margin of error but I think it's a low risk change that reduces code complexity and moves to the common ArrayPool that is heavily tested.

Drawaes commented:
/cc @geoffkizer @bartonjs @Priya91 @stephentoub

Drawaes commented:
@dotnet-bot Test Outerloop Windows x64 Debug Build
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot Test Outerloop Linux x64 Release Build

davidsh commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -22,9 +23,7 @@ internal class SslStreamInternal
         private static readonly AsyncProtocolCallback s_readFrameCallback = new AsyncProtocolCallback(ReadFrameCallback);
 
         private const int PinnableReadBufferSize = 4096 * 4 + 32;         // We read in 16K chunks + headers.
```
If you're removing PinnableBufferCache, then maybe these constants should be removed or renamed?

Drawaes replied:
I have made this change, ReadBufferSize.

Drawaes commented:
After an offline discussion to use the arraypool the move to make the write buffer per request rather than per connection is needed. This has a tiny bump in Unix perf (5k) but due to the lock convoy not many other changes. Due to the per request nature the write buffer can now be dynamic to the current chunk size + overhead which should fix #23040 

There should be a bump due to a reduction of allocations on windows.

davidsh commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -20,11 +21,10 @@ internal class SslStreamInternal
-        private const int PinnableWriteBufferSize = 4096 + 1024;        // We write in 4K chunks + encryption overhead.
-        private static PinnableBufferCache s_PinnableWriteBufferCache = new PinnableBufferCache("System.Net.SslStream", PinnableWriteBufferSize);
+        private const int FrameOverhead = 32;
+        private const int ReadBufferSize = 4096 * 4 + FrameOverhead;         // We read in 16K chunks + headers.
```
Is this constant used for writing as well?  There used to be 2 constants (read and write).  What happened to the constant for write?

Drawaes commented:
I made it disappear. The write buffer was a bit odd. It rented it and held it for the life time of the connection (bad for the generations in the buffer pool). And it was randomly sized... the max chunk size for linux is 32k the max chunk size for windows will be 16k.

It would rent that buffer and if it's not big enough or in use just allocate a non pooled buffer. That is what the issue is complaining about that I referenced.

So now it rents a buffer for a short time... e.g. encrypt to it and write and returns it. Because of this we know the size we need so we rent one of that size...

stephentoub commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -20,11 +21,10 @@ internal class SslStreamInternal
         private static readonly AsyncProtocolCallback s_resumeAsyncReadCallback = new AsyncProtocolCallback(ResumeAsyncReadCallback);
         private static readonly AsyncProtocolCallback s_readHeaderCallback = new AsyncProtocolCallback(ReadHeaderCallback);
         private static readonly AsyncProtocolCallback s_readFrameCallback = new AsyncProtocolCallback(ReadFrameCallback);
+        private static readonly Action<Task, object> s_freeWriteBufferCallback = new Action<Task, object>((_, o) => ArrayPool<byte>.Shared.Return((byte[])o));
```
The continuation needs to call .GetAwaiter().GetResult() on the passed in antecedent task in order to propagate any exception that may have occurred.

stephentoub commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -63,7 +55,7 @@ private void FreeReadBuffer()
         {
             if (_internalBufferFromPinnableCache)
```
Now that it's no longer the "pinnable cache", let's rename this to something else, maybe just "_internalBufferFromCache".  Same goes for related locals elsewhere, like "wasPinnable".

stephentoub commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -415,32 +360,37 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
                     }
-
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    catch
```
Other than the explicit throw a few lines up, what here might throw?  I'm wondering if we could just put the return just before the explicit throw above and get rid of the try/catch entirely.

Drawaes replied:
Yeah tracing through the code the only errors will be native and they are dealt with via error codes not exceptions

stephentoub commented:
Fixes https://github.com/dotnet/corefx/issues/15761

stephentoub commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -231,23 +209,13 @@ private void EnsureInternalBufferSize(int newSize)
                 byte[] saved = _internalBuffer;
 
-                if (newSize <= PinnableReadBufferSize)
+                if (newSize <= ReadBufferSize)
```
In what circumstance is newSize > ReadBufferSize?  And if it's common, why can't we use the pool for that, too? i.e. why is there a limit?

Drawaes replied:
None, and actually we end up with a 32k buffer often which we could better take advantage of (2^n for array pool buffers and as we are 2^14 and change we go to 2^15).  I honestly didn't massively look at the read side until last night as I have focused on the write, but there are some quick wins so I will fix these later today.

Drawaes commented:
Made all of @stephentoub changes. Removed a bunch of stuff not needed (isPinnable flag isn't needed as it is always pinnable). Added the freeing of the buffer sooner on the read side (if _internalBufferCount == 0 && _decryptedBytesCount == 0). 

Linux is the same, windows seems a little up (I suspect less allocations and removing a try/catch from the hot path).

Seeing about 900k/s on a Standard D15_v2 (20 Cores, 140 GB memory) on Azure

stephentoub commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -20,11 +21,10 @@ internal class SslStreamInternal
-        private const int PinnableWriteBufferSize = 4096 + 1024;        // We write in 4K chunks + encryption overhead.
-        private static PinnableBufferCache s_PinnableWriteBufferCache = new PinnableBufferCache("System.Net.SslStream", PinnableWriteBufferSize);
+        private const int FrameOverhead = 32;
+        private const int ReadBufferSize = 4096 * 4 + FrameOverhead;         // We read in 16K chunks + headers.
```
Where is ReadBufferSize used now?

Drawaes replied:
It's not but... I have just been prototyping exactly that. Because we say "min size 5" and fill that and then min size "frame" and then copy/move and carry on now. Interestingly it didn't hurt perf at all. But it causes a move back to the bad old days of reading 5 then frame.

I am thinking that whole ensure logic can go out the door and just use the readframe size. The TLS spec is pretty clear 14^2 for the plain text and then headers etc. TLS 1.3 is exactly the same.

So I say make it 14^2 + headers with a reference and ditch all the resizing. Just the reset to offset zero after a frame is used.

Drawaes replied:
And free of course if all data is consumed.

stephentoub commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -55,19 +55,30 @@ private static void FreeWriteBuffer(Task t, object buffer)
+                _decryptedBytesOffset = 0;
+                return true;
             }
+            return false;
```
Why does it return a bool if none of the callers use the bool?

Drawaes replied:
No good reason, just the Tryxxx convention but it didn't sit well with me either so I removed it.

stephentoub replied:
Doesn't need to be named `TryXxx`... could, for example, be something like `ReturnBufferIfEmpty`.

Drawaes replied:
I like that a lot better, renamed.

stephentoub commented:
@benaadams, @geoffkizer, any feedback?

benaadams commented:
LGTM

Drawaes commented:
Cool, onwards and upwards :)

geoffkizer commented:
Sorry for not chiming in here sooner, I've been ill.

@Drawaes, in the future, can we use the tests are https://github.com/geoffkizer/netperf/tree/master/SocketPerfTest to get data on these kinds of checkins?  The advantage of these it's that it's very focused on socket/SSL perf and excludes other variables.

Running on a big 20 core machine is nice and can help identify contention issues, but it's not representative of what most of our customers will do and it may hide other issues.  We should get data on smaller machines, e.g. 4 proc VMs.

I assume we're saturating the CPU when you're running these tests?

geoffkizer commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -104,8 +91,9 @@ internal int ReadByte()
-                    int b = InternalBuffer[_decryptedBytesOffset++];
+                    int b = _internalBuffer[_decryptedBytesOffset++];
                     _decryptedBytesCount--;
+                    ReturnReadBufferIfEmpty();
```
This seems like a significant change in behavior.  Why are we doing this in a change about removing PinnableBufferCache?  Do we really understand the impact of this change?

stephentoub replied:
Is this read buffer the same byte[] that's passed to NetworkStream under SslStream?  If so, there is the potential concern that we'd be forced to re-pin on every operation, if we end up getting back a different buffer from the pool, whereas if we hold onto the same buffer, the SocketAsyncEventArgs optimizations will kick in and we won't need to.

stephentoub replied:
@Drawaes, thinking about this more, I am concerned about the read buffer change.  Do any of the tests you've run validate perf of SslStream over a NetworkStream?

Drawaes replied:
I am not 100% sure about what you mean here? Am I understanding it correctly that you are saying

If I call
```
await stream.WriteAsync(someData);
// do some other stuff
await stream.WriteAsync(someData);
```

That from the first write to the second write socket has held a pin on my someData buffer?

stephentoub replied:
> That from the first write to the second write socket has held a pin on my someData buffer?

The way things are currently implemented... yes.

Drawaes replied:
So what happens if I call
` stream.WriteAsync(bla); `
Then don't call again for some prolonged period, is my buffer no longer eligible for GC and also causing memory fragmentation? (outside of the SslStream case here) ?

stephentoub replied:
> is my buffer no longer eligible for GC and also causing memory fragmentation

It's pinned, and everything that comes with it.

Drawaes replied:
It's actually interesting. Not sure I like the behaviour of the pin effects leaking, but putting that aside for a second and really unrelated to this PR. 

With buffers, you can pin them, I was thinking that an API for stream not just to pass in a buffer as is currently suggested, but where you could pass in a pin from the buffer as it has such a thing might provide a perf bump. So say SslStream could pin to interop and do the encrypt, and pass the pin to the underlying stream (network or otherwise) and release it after the operation has finished. This could just be ignored if the underlying stream doesn't support it with no harm but if it could use it then there could be a chance to reduce pin/free's all the way down the chain. 

Which btw pipelines avoids in a way by creating slabs of prepinned buffers.

stephentoub replied:
> I was thinking that an API for stream not just to pass in a buffer as is currently suggested, but where you could pass in a pin from the buffer as it has such a thing might provide a perf bump

With https://github.com/dotnet/corefx/issues/22608, you'll be able to give a `Buffer<T>` (now called `Memory<T>`) to SocketAsyncEventArgs.  Calling `Retain` on an already pinned buffer should be negligable cost.  If you want to use a pool of already pinned memory, you can.

stephentoub replied:
Regardless, let's circle back to my concern.  Have we validated the impact of this on SslStream over a NetworkStream?  Separately it's worthwhile re-evaluating the long-standing behavior of pinning in SocketAsyncEventArgs.SetBuffer so that individual send/receive operations don't have to, but that is the current behavior, and we should validate we're not regressing in the face of it.

Drawaes replied:
Agreed, I was unaware of that behaviour and will look into the impact.

stephentoub replied:
Thanks

geoffkizer commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -216,51 +204,11 @@ internal void EndWrite(IAsyncResult asyncResult)
-                }
+                _internalBuffer = ArrayPool<byte>.Shared.Rent(ReadBufferSize);
             }
             else if (_internalOffset > 0 && _internalBufferCount > 0)
```
Can we ever get into the state where _internalBuffer != null but one or both of these is 0?  What happens in that case?  At the very least, we should assert here.

Drawaes replied:
If internal buffer != null, then we have a rented buffer

If the offset is > 0 and the count > 0 then we need to shift the data to the front, 
if the count is zero and the offset is > 0 then we just reset the offset to zero
if the offset is zero and the count is > 0 then we have data but it's at the front anyway

So all of the above cases could/will happen and they are all dealt with, or am I missing something?

geoffkizer replied:
(1) When would we ever get into the case where count == 0 and internalBuffer != null?  Is that possible?  I don't think it is.  If not, we should assert it.  
(2) We don't seem to reset the _internalCount to 0 when we allocate a new buffer.  Perhaps this is guaranteed somewhere else?
(3) We should fix the name of the method, since it no longer has anything to do with size.

More than anything, I'm trying to understand the possible states we can get into here.  It wasn't obvious to me from the code.

Drawaes replied:
1. Yes an assert would be good for that, as it shouldn't happen anymore, it's due to me modifying this before the release method was baked in, it should never happen now.
2. Do you mean when we allocate a brand new buffer (eg buffer == null?) or when we allocate a new one and copy. In the first instance when we free it we set everything to zero, in the second case the copy should keep the same length.
3. Agreed

I will generally clean it up.

geoffkizer replied:
Re 2 -- I meant buffer == null.  If we're assuming it's always 0 when we reach here, we should assert that.  But personally I'd just set it to 0 here.

Drawaes replied:
I think the assert is needed because it should never happen due to
Either the buffer is null because we have never allocated... or it is null because of

```
private void ReturnReadBufferIfEmpty()
{
    if (_internalBuffer != null && _decryptedBytesCount == 0 && _internalBufferCount == 0)
    {
        ArrayPool<byte>.Shared.Return(_internalBuffer);
       _internalBuffer = null;
       _internalBufferCount = 0;
       _internalOffset = 0;
       _decryptedBytesCount = 0;
       _decryptedBytesOffset = 0;
    }
}
```

geoffkizer commented on src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs:
```diff
@@ -415,32 +340,32 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
                     int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
-                    int encryptedBytes;
+                    byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(chunkBytes + FrameOverhead);
+                    byte[] outBuffer = rentedBuffer;
```
outBuffer and rentedBuffer seem to always be the same.  Can we clean this up?

stephentoub replied:
A call below takes `outBuffer` by ref and may change it, and if it does we still need to be sure we're releasing the originally rented before to the pool, not whatever `outBuffer` was reassigned to by that ref.

benaadams replied:
`outBuffer` is passed by `ref` so isn't necessarily the same as `rentedBuffer` after the call?

geoffkizer replied:
Yeah, I missed the ref, thanks.

The fact that outBuffer is passed by ref is weird in and of itself.  Lower layers seem to do a null check and buffer size check, and allocate byte[] if the original buffer is null or too small.  This is almost certainly the wrong thing to do; we should just ensure that the originally allocated buffer is large enough.

That said, that's a problem for another day.

Drawaes replied:
Absolutely and before this change it was often the case, mostly because PinnableCache was fixed in size. This changes that but I didn't want to go through the layers and rip stuff out .... yet... as it touched all the PALs. I am trying to avoid massive sweeping changes in jumbo PR's :)

geoffkizer replied:
That's good :)

Drawaes commented:
Right so I used azure Windows 2016 4 cpu machines (2 of them in the same network with no firewall between them using local IP's)

I ran your tool, and let it run for a couple of minutes for each run and then took the average


Connections | Message Size | New | Original | Change
-- | -- | -- | -- | --
256 | 256 | 67441 | 66762 | 1.02%
256 | 4096 | 52558 | 53413 | -1.60%
256 | 32000 | 7264 | 7447 | -2.46%

So as the size increased it did slow down a bit. However the variation of the 3 sec recording would say they are pretty much all in the margin of error. If you are okay with that (for now) and with the idea that this will facilitate the future changes I will go ahead and make a PR to fix the issues that @geoffkizer has highlighted above?

geoffkizer commented:
I'm fine with the results as they are; we should move forward here.

I am a little surprised that larger sizes seem to see a loss, where smaller ones don't.  Not sure how to explain that exactly...

Drawaes commented:
Yeah I am not 100% sure, unless it is because we are allocating a lot of small buffers as the ArrayPool is dynamic sized, where as before we were moving around 5k buffers no matter what the write size is? It might also explain the small uptick in aspnet perf as it writes in smaller chunks in general.
